src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,199 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+
+namespace System.Data.Common
+{
+    public static partial class DbProviderFactories
+    {
+        private struct ProviderRegistration
+        {
+            internal ProviderRegistration(string factoryTypeAssemblyQualifiedName, DbProviderFactory factoryInstance)
+            {
+                ADP.CheckArgumentLength(factoryTypeAssemblyQualifiedName, nameof(factoryTypeAssemblyQualifiedName));
+                this.FactoryTypeAssemblyQualifiedName = factoryTypeAssemblyQualifiedName;
+                this.FactoryInstance = factoryInstance;
+            }
+
+            internal string FactoryTypeAssemblyQualifiedName { get; }
+            /// <summary>
+            /// The cached instance of the type in <see cref="FactoryTypeAssemblyQualifiedName"/>. If null, this registation is seen as a deferred registration
+            /// and <see cref="FactoryTypeAssemblyQualifiedName"/> is checked the first time when this registration is requested through GetFactory().
+            /// </summary>
+            internal DbProviderFactory FactoryInstance { get; }
+        }
+
+        private static ConcurrentDictionary<string, ProviderRegistration> _registeredFactories = new ConcurrentDictionary<string, ProviderRegistration>();
+        private const string AssemblyQualifiedName = "AssemblyQualifiedName";
+        private const string InvariantName = "InvariantName";
+        private const string Name = "Name";
+        private const string Description = "Description";
+        private const string ProviderGroup = "DbProviderFactories";
+        private const string Instance = "Instance";
+
+        public static bool TryGetFactory(string providerInvariantName, out DbProviderFactory factory)
+        {
+            factory = GetFactory(providerInvariantName, throwOnError: false);
+            return factory != null;
+        }
+
+        public static DbProviderFactory GetFactory(string providerInvariantName)
+        {
+            return GetFactory(providerInvariantName, throwOnError: true);
+        }
+
+        public static DbProviderFactory GetFactory(DataRow providerRow)
+        {
+            ADP.CheckArgumentNull(providerRow, nameof(providerRow));
+
+            DataColumn assemblyQualifiedNameColumn = providerRow.Table.Columns[AssemblyQualifiedName];
```

Comment:
IIRC @saurabh500 does not like var, but I don't like not using var when it is possible and clear, so I will call it out anyway :smile_cat: Not sure if there are general guidelines for new code in corefx. @karelz?

Reply:
Guidelines say afaik that 'var' is discouraged, but have to look it up.

Reply:
Guidelines here: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md point 10 says only about when not to use var, but it doesn't say use var when appropriate. There are many other cases in the class where var is an option. I didn't use that as the original code didn't and altering just for the sake of changing is also a bit much ;)